doc/designs/quic-design/congestion-control.md

```diff
@@ -0,0 +1,229 @@
+     */
+    int (*get_exemption)(OSSL_CC_DATA *ccdata);
+
+    /*
+     * Returns 1 if sending is allowed, 0 otherwise.
+     */
+    int (*can_send)(OSSL_CC_DATA *ccdata);
+
+    /*
+     * Returns number of bytes allowed to be sent.
+     * |time_since_last_send| is time since last send operation
+     * in microseconds.
+     * |time_valid| is 1 if the |time_since_last_send| is
+     * valid, 0 otherwise.
+     */
+    size_t (*get_send_allowance)(OSSL_CC_DATA *ccdata,
+                                 uint64_t time_since_last_send,
+                                 int time_valid);
+    /*
+     * Returns the maximum number of bytes allowed to be in flight.
+     */
+    size_t (*get_bytes_in_flight_max)(OSSL_CC_DATA *ccdata);
+
+    /*
+     * To be called when a packet with retransmittable data was sent.
+     * |num_retransmittable_bytes| is the number of bytes sent
+     * in the packet that are retransmittable.
+     * Returns 1 on success, 0 otherwise.
+     */
+    int (*on_data_sent)(OSSL_CC_DATA *ccdata,
+                        size_t num_retransmittable_bytes);
+
+    /*
+     * To be called when retransmittable data was invalidated.
+     * I.E. they are not considered in-flight anymore but
+     * are neither acknowledged nor lost. In particular used when
+     * 0RTT data was rejected.
+     * |num_retransmittable_bytes| is the number of bytes
+     * of the invalidated data.
+     * Returns 1 if sending is unblocked, 0 otherwise.
+     */
+    int (*on_data_invalidated)(OSSL_CC_DATA *ccdata,
+                               size_t num_retransmittable_bytes);
+
+    /*
+     * To be called when sent data was acked.
+     * |time_now| is current time in microseconds.
+     * |largest_pn_acked| is the largest packet number of the acked
+     * packets.
+     * |num_retransmittable_bytes| is the number of retransmittable
+     * packet bytes that were acked.
+     * |smoothed_rtt| is the smoothed round trip time of the
+     * connection in microseconds.
+     * Returns 1 if sending is unblocked, 0 otherwise.
+     */
+    int (*on_data_acked)(OSSL_CC_DATA *ccdata,
+                         uint64_t time_now,
+                         uint64_t last_pn_acked,
+                         size_t num_retransmittable_bytes,
+                         uint64_t smoothed_rtt);
+
+    /*
+     * To be called when sent data is considered lost.
+     * |largest_pn_lost| is the largest packet number of the lost
+     * packets.
+     * |largest_pn_sent| is the largest packet number of the sent
+     * packets.
+     * |num_retransmittable_bytes| is the number of retransmittable
+     * packet bytes that are considered lost.
+     * |persistent_congestion| is 1 if the congestion is considered
+     * persistent (i.e., the number of probes exceeded the persistent
+     * congestion threshold), 0 otherwise.
+     * Returns 1 if this is considered a new congestion event,
+     * 0 otherwise.
```

Comment:
I'm a little confused at how the abstraction boundary makes sense to leave determination of "persistent" external to the CC method but determination of "distinct congestion event" inside the method.